Tests are failing on Plan 9 because os.Rename doesn't rename across directory

<!--
Hello! To ensure this issue is correctly addressed as soon as possible by the IPFS team, please try to make sure:

- This issue is relevant to this repository's topic or codebase.

- A clear description is provided. It should includes as much relevant information as possible and clear scope for the issue to be actionable.

FOR GENERAL DISCUSSION, HELP OR QUESTIONS, please see the options at https://ipfs.io/help or head directly to https://discuss.ipfs.io.

(you can delete this section after reading)
-->
Example failure:
```
cpu% go test -run 'TestMove$'
...
--- FAIL: TestMove (4.24s)
    convert_test.go:216: Put fail: rename /tmp/test-datastore-flatfs-825411366/v1/.temp/temp-030917448 /tmp/test-datastore-flatfs-825411366/v1/XKC/XKCILQAO5ZLD2K===.data: invalid argument
```

--- a/convert.go
+++ b/convert.go
@@ -156,7 +156,7 @@ func Move(oldPath string, newPath string, out io.Writer) error {
 			// else we found something unexpected, so to be safe just move it
 			log.Warnw("found unexpected file in datastore directory, moving anyways", "file", fn)
 			newPath := filepath.Join(newDS.path, fn)
-			err := os.Rename(oldPath, newPath)
+			err := rename(oldPath, newPath)
 			if err != nil {
 				return err
 			}
@@ -177,7 +177,7 @@ func moveKey(oldDS *Datastore, newDS *Datastore, key datastore.Key) error {
 	if err != nil && !os.IsExist(err) {
 		return err
 	}
-	err = os.Rename(oldPath, newPath)
+	err = rename(oldPath, newPath)
 	if err != nil {
 		return err
 	}

--- a/flatfs.go
+++ b/flatfs.go
@@ -369,7 +369,7 @@ func (fs *Datastore) renameAndUpdateDiskUsage(tmpPath, path string) error {
 	// it will either a) Re-add the size of an existing file, which
 	// was sustracted before b) Add 0 if there is no existing file.
 	for i := 0; i < RetryAttempts; i++ {
-		err = os.Rename(tmpPath, path)
+		err = rename(tmpPath, path)
 		// if there's no error, or the source file doesn't exist, abort.
 		if err == nil || os.IsNotExist(err) {
 			break
@@ -1060,7 +1060,7 @@ func (fs *Datastore) writeDiskUsageFile(du int64, doSync bool) {
 	}
 	closed = true
 
-	if err := os.Rename(tmp.Name(), filepath.Join(fs.path, DiskUsageFile)); err != nil {
+	if err := rename(tmp.Name(), filepath.Join(fs.path, DiskUsageFile)); err != nil {
 		log.Warnw("cound not write disk usage", "error", err)
 		return
 	}

--- a/rename.go
+++ b/rename.go
@@ -1,0 +1,7 @@
+// +build !plan9
+
+package flatfs
+
+import "os"
+
+var rename = os.Rename

--- a/rename_plan9.go
+++ b/rename_plan9.go
@@ -1,0 +1,77 @@
+package flatfs
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// rename behaves like os.Rename but can rename files across directories.
+func rename(oldpath, newpath string) error {
+	err := os.Rename(oldpath, newpath)
+	if le, ok := err.(*os.LinkError); !ok || le.Err != os.ErrInvalid {
+		return err
+	}
+	if filepath.Dir(oldpath) == filepath.Dir(newpath) {
+		// We should not get here, but just in case
+		// os.ErrInvalid is used for something else in the future.
+		return err
+	}
+
+	src, err := os.Open(oldpath)
+	if err != nil {
+		return &os.LinkError{"rename", oldpath, newpath, err}
+	}
+	defer src.Close()
+
+	fi, err := src.Stat()
+	if err != nil {
+		return &os.LinkError{"rename", oldpath, newpath, err}
+	}
+	if fi.Mode().IsDir() {
+		return &os.LinkError{"rename", oldpath, newpath, syscall.EISDIR}
+	}
+
+	dst, err := os.OpenFile(newpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fi.Mode())
+	if err != nil {
+		return &os.LinkError{"rename", oldpath, newpath, err}
+	}
+
+	if _, err := io.Copy(dst, src); err != nil {
+		dst.Close()
+		os.Remove(newpath)
+		return &os.LinkError{"rename", oldpath, newpath, err}
+	}
+	if err := dst.Close(); err != nil {
+		os.Remove(newpath)
+		return &os.LinkError{"rename", oldpath, newpath, err}
+	}
+
+	// Copy mtime and mode from original file.
+	// We need only one syscall if we avoid os.Chmod and os.Chtimes.
+	dir := fi.Sys().(*syscall.Dir)
+	var d syscall.Dir
+	d.Null()
+	d.Mtime = dir.Mtime
+	d.Mode = dir.Mode
+	_ = dirwstat(newpath, &d) // ignore error, as per mv(1)
+
+	if err := os.Remove(oldpath); err != nil {
+		return &os.LinkError{"rename", oldpath, newpath, err}
+	}
+	return nil
+}
+
+func dirwstat(name string, d *syscall.Dir) error {
+	var buf [syscall.STATFIXLEN]byte
+
+	n, err := d.Marshal(buf[:])
+	if err != nil {
+		return &os.PathError{"dirwstat", name, err}
+	}
+	if err = syscall.Wstat(name, buf[:n]); err != nil {
+		return &os.PathError{"dirwstat", name, err}
+	}
+	return nil
+}

